metcdv3/balancer.go

```diff
@@ -0,0 +1,84 @@
+package metcdv3
+
+import (
+	"context"
+	"encoding/json"
+	"path"
+
+	etcdv3 "github.com/coreos/etcd/clientv3"
+	"github.com/lytics/metafora"
+)
+
+// NewFairBalancer creates a new metafora.DefaultFairBalancer that uses etcd
+// for counting tasks per node.
+func NewFairBalancer(conf *Config, etcdv3c *etcdv3.Client) metafora.Balancer {
+	e := etcdClusterState{
+		etcdv3c:  etcdv3c,
+		kvc:      etcdv3.NewKV(etcdv3c),
+		taskPath: path.Join(conf.Namespace, TasksPath),
+		nodePath: path.Join(conf.Namespace, NodesPath),
+	}
+	return metafora.NewDefaultFairBalancer(conf.Name, &e)
+}
+
+// Checks the current state of an Etcd cluster
+type etcdClusterState struct {
+	etcdv3c  *etcdv3.Client
+	kvc      etcdv3.KV
+	taskPath string
+	nodePath string
+}
+
+func (e *etcdClusterState) NodeTaskCount() (map[string]int, error) {
+	state := map[string]int{}
+
+	// First initialize state with nodes as keys
+	resp, err := e.kvc.Get(context.Background(), e.nodePath, etcdv3.WithPrefix())
+	if err != nil {
+		return nil, err
+	}
+
+	if resp == nil || len(resp.Kvs) == 0 {
+		metafora.Warnf("balancer received empty response from GET %s", e.nodePath)
+		return state, nil
+	}
+
+	for _, kv := range resp.Kvs {
+		// We're garunteed to find nodes under the _metadata path (created on Coordinator startup)
```

Comment:
typo :wink:

Reply:
The only thing we we're missing was u ;)